### ISSUE TITLE
Python console wrong character encoding

### ISSUE THREAD

**Issue by danzig666** [qgis/QGIS#43834]
**Describe the bug**
Python script character encoding is based on something else than UTF-8. Tried every Windows 10 setting to change region/locale/encoding with no success.

**How to Reproduce**

Enter print('áéŐŰÍ') in the python script window and run the script. I wonder where cp1252 is coming from. Entering it in the console command line works though.

![image](https://user-images.githubusercontent.com/5346511/122899193-e4d36800-d34b-11eb-8fe9-7efc50a3d27b.png)

**QGIS and OS versions**

QGIS version
3.20.0-Odense
QGIS code revision
decaadbb31
Qt version
5.15.2
Python version
3.9.5
GDAL/OGR version
3.3.0
PROJ version
8.0.1
EPSG Registry database version
v10.018 (2021-04-02)
GEOS version
3.9.1-CAPI-1.14.2
SQLite version
3.35.2
PDAL version
2.3.0
PostgreSQL client version
13.0
SpatiaLite version
5.0.1
QWT version
6.1.3
QScintilla2 version
2.11.5
OS version
Windows 10 Version 1809

Active Python plugins
db_manager
MetaSearch
processing

**Comment by gioman** [comment]
@danzig666 does it work correctly in previous releases?

**Comment by danzig666** [comment]
> @danzig666 does it work correctly in previous releases?

Yes it does. I use 3.18.3 currently, but it worked in all previous 3.x versions.

**Comment by nicogodet** [comment]
OK on 3.20 on Linux
NOK on 3.20 on Windows

**Comment by agiudiceandrea** [comment]
I can reproduce the issue on Windows with QGIS 3.18.1 from OSGeo4W MSI (new testing) installer.
I cannot reproduce the issue on Windows with QGIS 3.18.1 from OSGeo4W EXE (old regular) installer.

@danzig666, could you confirm QGIS 3.18.3 (with which you don't experience the issue) was installed using the OSGeo4W EXE installer?

**Comment by danzig666** [comment]
> @danzig666, could you confirm QGIS 3.18.3 (with which you don't experience the issue) was installed using the OSGeo4W EXE installer?

Yes, I installed QGIS 3.18.3 with the regular EXE installer (QGIS-OSGeo4W-3.18.3-1-Setup-x86_64.exe)

**Comment by agiudiceandrea** [comment]
So, IMHO this confirms the issue is related to the new Windows MSI installer (which ships a new Python version) and not by the new QGIS 3.20.0 version.

In fact, with the new Windows MSI installer the Python `locale` settings:
```python
import sys
bool(sys.flags.utf8_mode)
False

import locale
locale.getpreferredencoding(False)
'cp1252'
```

are different from the old Windows EXE installer:
```python
import sys
bool(sys.flags.utf8_mode)
True

import locale
locale.getpreferredencoding(False)
'UTF-8'
```

In console_editor.py `runScriptCode()` https://github.com/qgis/QGIS/blob/ea8d0d6e814444bd84fb99314c9b961ca38f79f8/python/console/console_editor.py#L459-L460
the `open` Python function is called without the `encoding` parameter, so it relies to the encoding returned by `locale.getpreferredencoding(False)` in order to properly decode the (temp) file and thus there is a different behavior between MSI and EXE installer.

A fix in QGIS code could be to explicitly enforce the utf-8 encoding:
```python
            self.parent.pc.shell.runCommand("exec(open('{0}'.encode('{1}'), encoding='utf-8').read())"
                                            .format(filename.replace("\\", "/"), sys.getfilesystemencoding()))
```
but I think the different `locale` behavior between new MSI and old EXE installers should also be addressed.

**Comment by gioman** [comment]
> So, IMHO this confirms the issue is related to the new Windows MSI installer (which ships a new Python version) and not by the new QGIS 3.20.0 version.

@agiudiceandrea as usual many thanks for the very detailed analysis.

**Comment by elpaso** [comment]
@agiudiceandrea maybe unrelated, but this rings me a bell: https://github.com/qgis/QGIS/pull/43249

**Comment by agiudiceandrea** [comment]
@elpaso maybe the same underlying cause. Do you experience the issue using QGIS from the new testing repository MSI installer? Could you test using the same QGIS version but from the old regular repository EXE installer or provide me a problematic py script and instruction to test on QGIS from both old and new repository?

**Comment by elpaso** [comment]
@agiudiceandrea sorry but I don't test windows regularly, that particular bug happened to a customer that was testing one of my Python plugins. The scenario was different though because in that case the form python code was loaded from a file and the problem was the default encoding of the loaded file which was set to win-cp-1252 instead of utf8.

**Comment by jef-n** [comment]
Try adding following to the ``etc\ini\python3-core.bat``:

``
set PYTHONUTF8=1
``

**Comment by nicogodet** [comment]
I add `PYTHONUTF8=1` to qgis-bin.env, .vars, etc\ini\python3.bat.
Just in case.

```python
>>> os.environ["PYTHONUTF8"]
'1'

>>> sys.getfilesystemencoding()
'utf-8'

>>> locale.getpreferredencoding()
'cp1252'
```

According to https://docs.python.org/3/using/cmdline.html#envvar-PYTHONUTF8, `locale.getpreferredencoding()` should return `utf-8`.

**Comment by agiudiceandrea** [comment]
@jef-n I confirm @nicogodet results setting `PYTHONUTF8=1` and using the Python console in QGIS (new repository), while using the OSGeo4W (new repository) console I could see different results:
```
run o-help for a list of available commands

F:\QGIS\SourceDir\qgis>python-qgis
Python 3.9.0 (tags/v3.9.0:9cf6752, Oct  5 2020, 15:34:40) [MSC v.1927 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.environ['PYTHONUTF8']
'1'
>>> import sys
>>> sys.flags.utf8_mode
1
>>> import locale
>>> locale.getpreferredencoding(False)
'UTF-8'
>>>
```

So it seems there's something preventing the setting to have the desired effect in QGIS.

Anyway I wonder why this setting should be needed now, since I don't find in the Python documentation a utf8_mode and locale.getpreferredencoding() any behaviour change between Python 3.7 and 3.9.

**Comment by nicogodet** [comment]
```diff
diff --git a/src/python/qgspythonutilsimpl.cpp b/src/python/qgspythonutilsimpl.cpp
index 18b7b46855..6d4a144d7a 100644
--- a/src/python/qgspythonutilsimpl.cpp
+++ b/src/python/qgspythonutilsimpl.cpp
@@ -163,6 +163,17 @@ bool QgsPythonUtilsImpl::checkSystemImports()
 void QgsPythonUtilsImpl::init()
 {
   // initialize python
+  PyStatus status;
+  PyPreConfig preconfig;
+  PyPreConfig_InitPythonConfig(&preconfig);
+
+  preconfig.utf8_mode = 1;
+
+  status = Py_PreInitialize(&preconfig);
+  if (PyStatus_Exception(status)) {
+      Py_ExitStatusException(status);
+  }
+
   Py_Initialize();
```

Using this, I force utf8_mode and `sys.flags.utf8_mode` returns 1. 

On linux, the same flag returned 0 before even with PYTHONUTF8=1 en env vars. Now, it returns 1.

**Comment by agiudiceandrea** [comment]
According to https://discuss.python.org/t/sys-flags-utf8-mode-is-1/431 and https://github.com/pyinstaller/pyinstaller/issues/3866 it seems that the facts that this issue (and https://github.com/qgis/QGIS/issues/41601 and https://github.com/qgis/QGIS/pull/43249 and probably other not yet reported) doesn't occur on Windows using QGIS shipped with the old OSGeo4W repository/installer was actually due to a bug in Python 3.7.0.

In Python 3.7.0 (shipped with the old OSGeo4W repository/installer), `sys.flags.utf8_mode` incorrectly returns `-1` also when PYTHONUTF8 was not set or not configured during initialization. `-1` is interpreted as `True` in the locale.getpreferredencoding control flow and so `locale.getpreferredencoding(False)` returns `UTF-8` on Windows.

In Python 3.9 (shipped with the new OSGeo4W repository/installer) the bug is not present (probably fixed in 3.7.1), so `sys.flags.utf8_mode` returns `0` when PYTHONUTF8 was not set or not configured during initialization and this lead `locale.getpreferredencoding(False)` return the 'ANSI' Windows current locale encoding.
On Linux locale.getpreferredencoding(False) usually always return `UTF-8` because usually the current locale encoding is `UTF-8`.

**Comment by dd-a** [comment]
Hi !
I have a very similar issue on Mac OS (10.13 and 10.14).
Lately I noticed that my Python scripts don't work anymore on QGIS, starting with version 3.12.  They all end with an error like ' UnicodeEncodeError: 'ascii' codec can't encode character '\xe9' in position 29: ordinal not in range(128) '. 
I encountered this with QGIS 3.12, 3.16.8 and 3.20.0
